plugins/pixart-tp/fu-pxi-tp-register.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright 2025 Harris Tai <harris_tai@pixart.com>
+ * Copyright 2025 Micky Hsieh <micky_hsieh@pixart.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include "fu-pxi-tp-register.h"
+
+#define REPORT_ID_SINGLE 0x42
+#define REPORT_ID_BURST	 0x41
+#define REPORT_ID_USER	 0x43
+#define OP_READ		 0x10
+
+gboolean
+fu_pxi_tp_register_write(FuPxiTpDevice *self, guint8 bank, guint8 addr, guint8 val, GError **error)
+{
+	guint8 buf[4] = {REPORT_ID_SINGLE, addr, bank, val};
+	g_autoptr(GError) error_local = NULL;
+
+	g_return_val_if_fail(FU_IS_PXI_TP_DEVICE(self), FALSE);
+
+	if (!fu_pxi_tp_common_send_feature(self, buf, sizeof(buf), &error_local)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "register write failed: bank=0x%02x addr=0x%02x val=0x%02x: %s",
+			    bank,
+			    addr,
+			    val,
+			    error_local != NULL ? error_local->message : "no details");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_pxi_tp_register_read(FuPxiTpDevice *self,
+			guint8 bank,
+			guint8 addr,
+			guint8 *out_val,
+			GError **error)
+{
+	guint8 cmd[4] = {REPORT_ID_SINGLE, addr, (guint8)(bank | OP_READ), 0x00};
+	guint8 resp[4] = {REPORT_ID_SINGLE};
+	g_autoptr(GError) error_local = NULL;
+
+	g_return_val_if_fail(FU_IS_PXI_TP_DEVICE(self), FALSE);
+	g_return_val_if_fail(out_val != NULL, FALSE);
+
+	if (!fu_pxi_tp_common_send_feature(self, cmd, sizeof(cmd), &error_local)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "register read command failed: bank=0x%02x addr=0x%02x: %s",
+			    bank,
+			    addr,
+			    error_local != NULL ? error_local->message : "no details");
+		return FALSE;
+	}
+
+	error_local = NULL;
```

Comment:
this is a bit of an antipattern -- you don't need error_local here at all. Just do:

	if (!fu_pxi_tp_common_send_feature(self, cmd, sizeof(cmd), error)) {
		g_prefix_error(error,
			    "register read command failed: bank=0x%02x addr=0x%02x: ",
			    bank,
			    addr);
		return FALSE;
	}